File: ios/DesMoinesInsider/Models/Attraction.swift
import Foundation
import CoreLocation

struct Attraction: Identifiable, Codable, Hashable {
    let id: String
    let name: String
    let type: String
    var location: String?
    var description: String?
    var rating: Double?
    var website: String?
    var imageUrl: String?
    var isFeatured: Bool?
    var latitude: Double?
    var longitude: Double?
    var createdAt: String?
    var updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id, name, type, location, description, rating, website, latitude, longitude
        case imageUrl = "image_url"
        case isFeatured = "is_featured"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    // MARK: - Computed Properties

    var attractionType: AttractionType {
        AttractionType(rawValue: type) ?? .other
    }

    var coordinate: CLLocationCoordinate2D? {
        // A missing coordinate is null in the DB (Double? == nil); a literal 0.0
        // is a valid location and must not be masked (IOS-AUDIT-BUG-016).
        guard let lat = latitude, let lng = longitude else { return nil }
        return CLLocationCoordinate2D(latitude: lat, longitude: lng)
    }

    /// Safe http/https website URL only (IOS-AUDIT-SEC-002) — an unsafe scheme
    /// in the content row yields nil so no button renders.
    var websiteURL: URL? {
        website.flatMap { $0.safeWebURL }
    }

    var ratingText: String {
        guard let rating else { return "No rating" }
        return String(format: "%.1f", rating)
    }

    /// Full VoiceOver label for a compact card button (name, type, rating).
    /// Mirrors `Restaurant.compactCardAccessibilityLabel` so the home rails
    /// read consistently (IOS-IA-001).
    var compactCardAccessibilityLabel: String {
        var parts: [String] = [name, attractionType.displayName]
        if rating != nil { parts.append("Rated \(ratingText)") }
        return parts.joined(separator: ". ")
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: Attraction, rhs: Attraction) -> Bool {
        lhs.id == rhs.id
    }
}

extension Attraction {
    static let preview = Attraction(
        id: "preview-1",
        name: "Pappajohn Sculpture Park",
        type: "Park",
        location: "1330 Grand Ave, Des Moines",
        description: "A 4.4-acre park featuring 30+ world-renowned sculptures by internationally acclaimed artists.",
        rating: 4.8,
        website: "https://desmoinesartcenter.org",
        isFeatured: true,
        latitude: 41.5862,
        longitude: -93.6354
    )
}

File: android/app/src/main/java/com/desmoines/aipulse/data/model/Attraction.kt
package com.desmoines.aipulse.data.model

import android.net.Uri
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.util.Locale

/**
 * Attraction data model matching the Supabase `attractions` table.
 * Mirrors iOS Attraction.swift.
 */
@Serializable
data class Attraction(
    val id: String,
    val name: String,
    val type: String,
    val location: String? = null,
    val description: String? = null,
    val rating: Double? = null,
    val website: String? = null,
    @SerialName("image_url") val imageUrl: String? = null,
    @SerialName("is_featured") val isFeatured: Boolean? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("updated_at") val updatedAt: String? = null,
) {

    /** Parsed AttractionType from the type string. */
    val attractionType: AttractionType
        get() = AttractionType.from(type)

    val coordinate: LatLng?
        get() {
            val lat = latitude ?: return null
            val lng = longitude ?: return null
            if (lat == 0.0 && lng == 0.0) return null
            return LatLng(lat, lng)
        }

    val websiteUri: Uri?
        get() {
            if (website.isNullOrEmpty()) return null
            val url = if (website.startsWith("http")) website else "https://$website"
            return Uri.parse(url)
        }

    val ratingText: String
        get() = rating?.let { String.format(Locale.US, "%.1f", it) } ?: "No rating"

    companion object {
        val preview = Attraction(
            id = "preview-1",
            name = "Pappajohn Sculpture Park",
            type = "Park",
            location = "1330 Grand Ave, Des Moines",
            description = "A 4.4-acre park featuring 30+ world-renowned sculptures by internationally acclaimed artists.",
            rating = 4.8,
            website = "https://desmoinesartcenter.org",
            isFeatured = true,
            latitude = 41.5862,
            longitude = -93.6354,
        )
    }
}
